### ISSUE TITLE
Parallel QuickCheck test execution

### ISSUE THREAD

**Issue by 0xd34df00d** [hspec/hspec#433]
Hspec is great, and the `parallel` combinator is particularly awesome! Unfortunately, there's a tiny shortcoming with it: the examples in each individual QuickCheck property test still get to run sequentially. For instance, given
```haskell
spec = parallel $
  describe "Some QuickCheck stuff $
    it "foo prop holds" $ property $ ...
```
then all the examples for this property are verified without any parallelism.

So, is it possible to parallelize their execution?

The closest I managed to find is [this](http://hackage.haskell.org/package/pqc-0.8), but I have no idea how to integrate that with Hspec.

If you think it's something that could be added somewhat easily, I'd be happy to hack on this given some pointers where to start looking.

**Comment by sol** [comment]
> the examples in each individual QuickCheck property test still get to run sequentially

How did you come to that conclusion?

Have you looked at http://hspec.github.io/parallel-spec-execution.html?

Note that even with parallelism, the test output will still be sequential, but the actual tests will evaluate is parallel.

**Comment by sol** [comment]
Also take a look at the `--jobs` option: http://hspec.github.io/options.html

**Comment by sol** [comment]
Here is an example that you can use to proof to yourself that tests are run in parallel:

```haskell
import           Test.Hspec
import           Test.QuickCheck
import           Control.Concurrent

main :: IO ()
main = hspec spec

spec :: Spec
spec = parallel $
  describe "Some QuickCheck stuff" $ do
    it "" $ property $ \ n -> do
        let y = n :: Int
        threadDelay 0
        putStrLn "XXXXXXXXXX"

    it "" $ property $ \ n -> do
        let y = n :: Int
        threadDelay 0
        putStrLn "YYYYYYYYY"

    it "" $ property $ \ n -> do
        let y = n :: Int
        threadDelay 0
        putStrLn "ZZZZZZZZ"
```

```
$ ghci Spec.hs
*Main> :main -j4
```

**Comment by 0xd34df00d** [comment]
> How did you come to that conclusion?

Empirically, sort of. Firstly, the time to run 100 QuickCheck tests is 100 times more than it takes to run a single test.

Then, to get a better proof, assume I have a `Spec` that roughly goes as follows:
```haskell
spec :: Spec
spec = parallel $ beforeAll acquireResource $ afterAll releaseResource $ do
  describe "Basic smoke tests" $ do
    it "baz 0" $ runTest "Input 0"
    describe "bar" $ do
      it "baz 1" $ runTest "Input 1"
      it "baz 2" $ runTest "Input 2"
      it "baz 3" $ runTest "Input 3"
  describe "QuickCheck fun" $
    it "arbitrary baz" $ \ih -> property $ \input -> runTest input
```

Given this, if I add a `putStrLn "starting"` and `putStrLn "stopping"` to the beginning and end of the `runTest` function (which is roughly an IO action), I observe them interleaving for the `"Basic smoke tests"` group, but `"QuickCheck fun"` group goes as
```haskell
QuickCheck fun
0/100
1/100"starting"
"stopping"
2/100"starting"
"stopping"
3/100"starting"
"stopping"
4/100"starting"
"stopping"
```
which I think clearly shows that a given property is tested sequentially.

> Here is an example that you can use to proof to yourself that tests are run in parallel

This is slightly different, having multiple QuickCheck properties indeed gets them tested in parallel, but each one of them still runs sequentially. Sorry I wasn't clear in my first comment.

**Comment by sol** [comment]
@0xd34df00d your mentioned `pqc` earlier.  Are you sure that `pqc` does something fundamentally different here?  From looking at the code I would assume that individual properties also run sequentially.

**Comment by 0xd34df00d** [comment]
@sol 
> Are you sure that pqc does something fundamentally different here? 

Nope, I'm not by any means. In fact I've just glanced over the documentation for it, and the docs for `pRunWithNum` seem to imply that it does.

Again, I haven't tried it myself, though — it's just the closest I was able to find from a very quick googling.

**Comment by sol** [comment]
As I see it, what you are asking for is currently not supported by `QuickCheck`.  If QuickCheck will support this is the feature then Hspec will benefit from it, likely without any changes to Hspec itself.

I'm closing this for now.

**Comment by sol** [comment]
@0xd34df00d I'm not sure if it is feasible to add this to QuickCheck, but you may want to consider to open a feature request for QuickCheck.

**Comment by Ericson2314** [comment]
Did you ever make progress with this?
